src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,401 @@
+           entrypointName=entrypointName, callContext] (Worker::Lock& lock) mutable ->
+              kj::Promise<void> {
+
+        jsg::Lock& js = lock;
+        if (!FeatureFlags::get(js).getJsRpc()) {
+          // JS RPC is not enabled on the server side, we cannot call any methods.
+          auto rpcDisabled =
+              kj::str("The receiving Worker does not allow its methods to be called over RPC."_kj);
+          // We need to save the JsValue here, otherwise the 'windows-debug' build complains.
+          auto err = js.error(rpcDisabled);
+          auto serialized = serializeV8(js, err);
+          auto builder = callContext.initResults(
+              capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+          builder.setV8Serialized(kj::mv(serialized));
+          co_return;
+        }
+
+        // We will try to get the function, if we can't we'll throw an error to the client.
+        auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+        KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+          auto serialized = serializeV8(js, js.error(e));
+          auto builder = callContext.initResults(
+              capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+          builder.setV8Serialized(kj::mv(serialized));
+          co_return;
+        }
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+
+        // We have a function, so let's call it and serialize the result for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        // If the function returns a promise we will wait for the promise to finish so we can
+        // serialize the result.
+        auto result = co_await callFuncAndSerialize(js, fn, serializedArgs);
+
+        auto builder = callContext.initResults(capnp::MessageSize { result.data.size() / 8 + 8, 0 });
+        if (result.caughtException) {
+          builder.initException().setV8Serialized(kj::mv(result.data));
+        } else {
+          builder.initValue().setV8Serialized(kj::mv(result.data));
         }
-    ));
+      }));
+    } catch(kj::Exception& e) {
+      if (auto desc = e.getDescription();
+          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
+        LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
+      }
+    }
+    // Upon returning, we want to fulfill the callPromise so customEvent can continue executing.
+    KJ_DEFER(callFulfiller->fulfill(););
+    co_return;
+  }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // Prevents us from calling a function defined on the Object Prototype, or a registered method.
+  bool verifyCallable(jsg::Lock& js, v8::Local<v8::Object> workerObj, kj::StringPtr name) {
+    auto workerProto = workerObj->GetPrototype().As<v8::Object>();
+    auto objectProto = workerProto->GetPrototype().As<v8::Object>();
```

Comment:
I think this isn't correct in the case of an inheritance heirarchy...

```
class Foo {}
class Bar extends Foo {}
```

In this case, Bar's prototype's prototype is Foo's prototype, not Object's prototype.

Reply:
And I think that we probably should allow methods of an inherited class (other than Object) to be invoked over RPC.

The only reason we don't want to expose `Object`'s methods is because the application has no control over them. But it is the application's choice to inherit from any other class.

Reply:
https://github.com/cloudflare/workerd/compare/f12f95f6a9e2c4ce0fd51d7c21e54c46276ae140..2a76ed3adeb7fd5f71fff31241b156573e06f074

Iterating until the end of the prototype chain now.